module/zstd/zstd.c

```diff
@@ -0,0 +1,662 @@
+		mutex_enter(&zstd_dctx_fallback.barrier);
+		z = zstd_dctx_fallback.mem;
+		type = ZSTD_KMEM_DCTX;
+	}
+
+	/* allocation should always be successful */
+	if (!z) {
+		return (NULL);
+	}
+
+	z->kmem_type = type;
+	z->kmem_size = nbytes;
+
+	return ((void*)z + (sizeof (struct zstd_kmem)));
+}
+
+static void
+zstd_free(void *opaque __unused, void *ptr)
+{
+	struct zstd_kmem *z = (ptr - sizeof (struct zstd_kmem));
+
+	ASSERT3U(z->kmem_type, <, ZSTD_KMEM_COUNT);
+	ASSERT3U(z->kmem_type, >=, ZSTD_KMEM_UNKNOWN);
+
+	if (((enum zstd_kmem_type)z->kmem_type) == ZSTD_KMEM_UNKNOWN) {
+		if (z->pool) {
+			zstd_mempool_free(z);
+		} else {
+			kmem_free(z, z->kmem_size);
+		}
+	} else {
+		mutex_exit(&zstd_dctx_fallback.barrier);
+	}
+}
+
+static void __init
+create_fallback_mem(struct zstd_fallback_mem *mem, size_t size)
+{
+	mem->mem_size = size;
+	mem->mem = vmem_zalloc(mem->mem_size, KM_SLEEP);
+	mutex_init(&mem->barrier, NULL, MUTEX_DEFAULT, NULL);
+}
+
+/* Initialize memory pool barrier mutexes */
+static void __init
+zstd_mempool_init(void)
+{
+	int i;
+
+	zstd_mempool_cctx = (struct zstd_pool *)
+	    kmem_zalloc(ZSTD_POOL_MAX * sizeof (struct zstd_pool), KM_SLEEP);
+	zstd_mempool_dctx = (struct zstd_pool *)
+	    kmem_zalloc(ZSTD_POOL_MAX * sizeof (struct zstd_pool), KM_SLEEP);
+
+	for (i = 0; i < ZSTD_POOL_MAX; i++) {
+		mutex_init(&zstd_mempool_cctx[i].barrier,
+		    NULL, MUTEX_DEFAULT, NULL);
+		mutex_init(&zstd_mempool_dctx[i].barrier,
+		    NULL, MUTEX_DEFAULT, NULL);
+	}
+}
+
+static int __init
+zstd_meminit(void)
+{
+	zstd_mempool_init();
+
+	/* Estimate the size of the fallback decompression context */
+	create_fallback_mem(&zstd_dctx_fallback,
+	    P2ROUNDUP(ZSTD_estimateDCtxSize() + sizeof (struct zstd_kmem),
+	    PAGESIZE));
+
+	return (0);
+}
+
+/* Release object from pool and free memory */
+static void __exit
+release_pool(struct zstd_pool *pool)
+{
+	mutex_enter(&pool->barrier);
+	mutex_exit(&pool->barrier);
+	mutex_destroy(&pool->barrier);
+	kmem_free(pool->mem, pool->size);
+	pool->mem = NULL;
+	pool->size = 0;
+}
+
+/* Release memory pool objects */
+static void __exit
+zstd_mempool_deinit(void)
+{
+	int i;
+
+	for (i = 0; i < ZSTD_POOL_MAX; i++) {
+		release_pool(&zstd_mempool_cctx[i]);
+		release_pool(&zstd_mempool_dctx[i]);
+	}
+
+	kmem_free(zstd_mempool_dctx,
+	    ZSTD_POOL_MAX * sizeof (struct zstd_pool));
+	kmem_free(zstd_mempool_cctx,
+	    ZSTD_POOL_MAX * sizeof (struct zstd_pool));
+}
+
+int __init
+zstd_init(void)
+{
+	/* Set pool size by using maximum sane thread count * 4 */
+	pool_count = boot_ncpus * 4;
+	zstd_meminit();
+	return (0);
+}
+
+void __exit
+zstd_fini(void)
+{
+	kmem_free(zstd_dctx_fallback.mem, zstd_dctx_fallback.mem_size);
+	mutex_destroy(&zstd_dctx_fallback.barrier);
+	zstd_mempool_deinit();
+}
+
+#if defined(_KERNEL)
+module_init(zstd_init);
+module_exit(zstd_fini);
+
+ZFS_MODULE_DESCRIPTION("ZSTD Compression for ZFS");
+ZFS_MODULE_AUTHOR(ZFS_META_AUTHOR);
+ZFS_MODULE_LICENSE(ZFS_META_LICENSE);
```

Comment:
Try something like:
ZFS_MODULE_LICENSE("Dual BSD/GPL");

(I think it's clear this file itself is CDDL)
See:
https://github.com/zfsonlinux/zfs/pull/8044#discussion_r227134618

Reply:
Well, this was because it "somehow" became GPL, while it actually *is* CDDL

Reply:
I was not talking about the copyright change, this line should refer to the ZSTD lib licence.

Reply:
mmh thats complicated. we have zstd which was licensed as BSD by Allan and zstd has its own license. now how to combine this if linked together? another option. doing another .ko module just for zstd library which exports the functions we need. but thats a bad solution

Reply:
thx, fixed again

Reply:
@Brainslayer
ZSTD is licenced CDDL, you relicenced it BSD without consent.
I know because I just read through that discussion, you asked permission to relicence and you did NOT get it.

Anyway, according to Behlendorf this should point to the licence of the module. As we already use the version of ZSTD, it's pretty clear that this should point to the licence for ZSTD. (which is a dual bsd/gpl licence).

The files outside modules (our connection module for example) are individually labeled anyway, its just to make clear that the library isn't the same)

Reply:
**Resolved**